token/rust/src/token.rs

```diff
@@ -0,0 +1,260 @@
+use super::client::{SendTransaction, TokenClient, TokenClientError};
+use solana_sdk::{
+    instruction::Instruction,
+    program_error::ProgramError,
+    program_pack::Pack,
+    pubkey::Pubkey,
+    signer::{signers::Signers, Signer},
+    system_instruction,
+    transaction::Transaction,
+};
+use spl_associated_token_account::{create_associated_token_account, get_associated_token_address};
+use spl_token::{instruction, state};
+use std::{fmt, sync::Arc};
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum TokenError {
+    #[error("client error: {0}")]
+    Client(TokenClientError),
+    #[error("program error: {0}")]
+    Program(#[from] ProgramError),
+    #[error("account not found")]
+    AccountNotFound,
+    #[error("invalid account owner")]
+    AccountInvalidOwner,
+    #[error("invalid account mint")]
+    AccountInvalidMint,
+}
+
+pub type TokenResult<T> = Result<T, TokenError>;
+
+pub struct Token<ST, TS> {
```

Comment:
Heh, `<ST, TS>` is a bit much. How about just <T, S>?
T: type / Transaction (sender)
S: Signer